chore(auth): add js passwordless changes (#8129)
aws-amplify/docs: src/pages/[platform]/build-a-backend/auth/connect-your-frontend/sign-in/index.mdx

| `CONFIRM_SIGN_IN_WITH_TOTP_CODE` | The sign-in must be confirmed with a TOTP code from the user. Complete the process with `confirmSignIn`. |
| `CONFIRM_SIGN_IN_WITH_SMS_CODE` | The sign-in must be confirmed with a SMS code from the user. Complete the process with `confirmSignIn`. |
| `CONFIRM_SIGN_IN_WITH_EMAIL_CODE` | The sign-in must be confirmed with a EMAIL code from the user. Complete the process with `confirmSignIn`. |

> | `CONFIRM_SIGN_IN_WITH_PASSWORD` | The sign-in must be confirmed with the password from the user. Complete the process with `confirmSignIn`. |

Author: @jjarvisp, is there a case where `CONTINUE_SIGN_IN_WITH_FIRST_FACTOR_SELECTION` could be the result of a `confirmSignIn` call?

Reviewer: Yes, `confirmSignIn()` calls `RespondToAuthChallenge` which could potentially return a `SELECT_CHALLENGE` challenge.

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_RespondToAuthChallenge.html#CognitoUserPools-RespondToAuthChallenge-response-ChallengeName

In order to facilitate the new passwordless sign in options, Cognito is introducing a new auth flow type known as `USER_AUTH`. This flow is designed to be flexible and supports both password and passwordless sign in factors.

```ts
const { nextStep } = await signIn({
username: state.username,
   options: {
        authFlowType: "USER_AUTH",
   }
});
```

### Sign In with a Preferred Challenge for First Factor

The USER_AUTH sign in flow will support the following methods of first factor authentication: `WEB_AUTHN`, `EMAIL_OTP`, `SMS_OTP`, `PASSWORD`, and `PASSWORD_SRP`.

If the desired first factor is known before the sign in flow is initiated it can be passed to the initial sign in call:

```ts
type AuthFactorType = 
    | 'WEB_AUTHN' 
    | 'EMAIL_OTP' 
    | 'SMS_OTP' 
    | 'PASSWORD' 
    | 'PASSWORD_SRP';

type SignInOptions = AuthServiceOptions & {
    authFlowType?: AuthFlowType;
    clientMetadata?: ClientMetadata;
    preferredChallenge?: AuthFactorType;
};
```

```ts
// PASSWORD_SRP / PASSWORD
// sign in with preferred challenge as password
// note password must be provided in same step
const { nextStep } = await signIn({
    username: state.username,

>     password: state.password,

Reviewer 2: can we use placeholder strings for these?

Author: Do you mean a variable with a placeholder string or string literals with a consistent placeholder string?

    },
});

// nextStep.signInStep === 'DONE'
handleNextSignInStep(nextStep);

// WEB_AUTHN
// sign in with preferred challenge as web authn
// no user input required at this step 
const { nextStep } = await signIn({
    username: state.username,
    options: {
        authFlowType: "USER_AUTH",
        preferredChallenge: "WEB_AUTHN" 
    },
});

// nextStep.signInStep === 'DONE'
handleNextSignInStep(nextStep);

// EMAIL_OTP
// sign in with preferred challenge as email otp
// no user input required at this step 
const { nextStep } = await signIn({
    username: state.username,
    options: {
        authFlowType: "USER_AUTH",
        preferredChallenge: "EMAIL_OTP" 
    },
});

// nextStep.signInStep === 'CONFIRM_SIGN_IN_WITH_EMAIL_CODE'
handleNextSignInStep(nextStep);

// SMS_OTP
// sign in with preferred challenge as sms otp
// no user input required at this step 
const { nextStep } = await signIn({
    username: state.username,
    options: {
        authFlowType: "USER_AUTH",
        preferredChallenge: "SMS_OTP" 
    },
});

// nextStep.signInStep === 'CONFIRM_SIGN_IN_WITH_SMS_CODE'
handleNextSignInStep(nextStep);
```

> Note that if the preferred auth factor is not available, the preference will be ignored and the flow will continue to first factor discovery.

Reviewer: I've seen this before 🤔

Your application's users can also sign in using passwordless methods. To learn more, visit the [concepts page for passwordless](/[platform]/build-a-backend/auth/concepts/passwordless/)

<InlineFilter filters={["angular", "javascript", "nextjs", "react", "react-native", "vue"]}>

### New auth flow type

In order to facilitate the new passwordless sign in options, Cognito is introducing a new auth flow type known as `USER_AUTH`. This flow is designed to be flexible and supports both password and passwordless sign in factors.

```ts
const { nextStep } = await signIn({

> username: state.username,

Reviewer: Indentation is off here.

Reviewer: Also, this info should be on the switching authentication flows page I believe.

    options: {
        authFlowType: "USER_AUTH",
    }
});
```

> ### Sign In with a Preferred Challenge for First Factor

Reviewer: Also, be careful with headers, we created 3 new sections on this page. Should only be SMS OTP, Email OTP, WebAuthn Passkeys, and Password or SRP.

Author: Are we restricted to those 3? I think giving details on the new `USER_AUTH` flow is a necessary prerequisite to handle the new passwordless methods.

Reviewer: We want to be consistent with the other libraries. We'd provide conceptual introduction to the auth flow on the `Switching Authentication Flows` page.

Reviewer: A good compromise might be to add a `<Callout />` mentioning `USER_AUTH` with a link to that page?

> The USER_AUTH sign in flow will support the following methods of first factor authentication: `WEB_AUTHN`, `EMAIL_OTP`, `SMS_OTP`, `PASSWORD`, and `PASSWORD_SRP`.

Reviewer: Keep the formatting consistent -> `USER_AUTH` instead of USER_AUTH

If the desired first factor is known before the sign in flow is initiated it can be passed to the initial sign in call:

```ts
type AuthFactorType = 
    | 'WEB_AUTHN' 
    | 'EMAIL_OTP' 
    | 'SMS_OTP' 
    | 'PASSWORD' 
    | 'PASSWORD_SRP';

type SignInOptions = AuthServiceOptions & {
    authFlowType?: AuthFlowType;
    clientMetadata?: ClientMetadata;
    preferredChallenge?: AuthFactorType;

> };

Reviewer: It might be confusing to have types here that aren't referenced in the examples; the connection between these types and how to pass as options is unclear. Is there a way we can clarify this? Maybe using the `SignInInput` interface which is publicly exposed by the library... or just relying on the examples?

```

```ts
// PASSWORD_SRP / PASSWORD
// sign in with preferred challenge as password
// note password must be provided in same step
const { nextStep } = await signIn({
    username: "hello@mycompany.com",
    password: "hunter2",
    options: {
        authFlowType: "USER_AUTH",
        preferredChallenge: "PASSWORD_SRP" // or "PASSWORD"
    },
});

> // nextStep.signInStep === 'DONE'

Reviewer: `nextStep` may not be DONE in all cases. We should provide a partial implementation of the `handleNextSignInStep()` handler

// SMS_OTP
// sign in with preferred challenge as sms otp
// no user input required at this step 
const { nextStep } = await signIn({
    username: "hello@mycompany.com",
    options: {
        authFlowType: "USER_AUTH",
        preferredChallenge: "SMS_OTP" 
    },
});

// nextStep.signInStep === 'CONFIRM_SIGN_IN_WITH_SMS_CODE'
handleNextSignInStep(nextStep);
```

Note that if the preferred auth factor is not available, the preference will be ignored and the flow will continue to first factor discovery.

### Passwordless Sign In with First Factor Discovery

When multiple first factor options are available and none are set as preferred in InitiateAuth API call (or preferred challenge is not available), Amplify JS continues to a new sign in step `CONTINUE_SIGN_IN_WITH_FIRST_FACTOR_SELECTION`:

```ts
interface ContinueSignInWithFirstFactorSelection {
    signInStep: 'CONTINUE_SIGN_IN_WITH_FIRST_FACTOR_SELECTION';
    availableChallenges?: ChallengeName[];
}
```

```ts
const { nextStep } = await signIn({
    username: "hello@mycompany.com",
    options: {
        authFlowType: "USER_AUTH",
    }
});

// nextStep.signInStep === 'CONTINUE_SIGN_IN_WITH_FIRST_FACTOR_SELECTION'

> handleNextSignInStep(nextStep);

Reviewer: I think we should avoid including internal types in the documentation; we should show an example of how to use the availableChallenges instead.

```

> To initiate an available first factor, the desired first factor type is passed to confirm sign in as the challenge response. The next sections will that step for each `AuthFactorType`.

Reviewer: Not sure what's trying to be said here, maybe something like below?
```suggestion
To initiate an available first factor, the desired first factor type is passed to confirm sign in as the challenge response. The next sections will cover how to handle each challenge during the passwordless authentication flow.
```

<InlineFilter filters={["angular", "javascript", "nextjs", "react", "react-native", "vue"]}>

To request an OTP code via email for authentication, the challenge is passed as the challenge response to the confirm sign in API.

> Amplify JS will respond appropriately to Cognito and return the challenge as sign in next step: `CONFIRM_SIGN_IN_WITH_EMAIL_CODE`:

Reviewer: ```suggestion
Amplify will respond appropriately to Cognito and return the challenge as sign in next step: `CONFIRM_SIGN_IN_WITH_EMAIL_CODE`:
```

<InlineFilter filters={["angular", "javascript", "nextjs", "react", "react-native", "vue"]}>

> The WebAuthn credential flow is initiated by passing the challenge name to the confirm sign in api. Cognito will respond with credential request options that are used by Amplify JS to begin the authentication ceremony with the local authenticator.

Reviewer: ```suggestion
The WebAuthn credential flow is initiated by passing the challenge name to the confirm sign in api. Cognito will respond with credential request options that are used by Amplify to begin the authentication ceremony with the local authenticator.
```

> As no additional user input is required from the application to complete this flow, there is no intermediate sign in step specific to WebAuthn:

Reviewer: May not need to call this out, as it's more of an implementation detail.
